streams/src/main/java/org/apache/kafka/streams/processor/internals/namedtopology/KafkaStreamsNamedTopologyWrapper.java

```diff
@@ -230,53 +233,72 @@ public RemoveNamedTopologyResult removeNamedTopology(final String topologyToRemo
 
         topologyMetadata.unregisterTopology(removeTopologyFuture, topologyToRemove);
 
-        if (resetOffsets) {
+        if (!completedFutureForUnstartedApp(removeTopologyFuture, "removing topology") && resetOffsets) {
```

Comment:
nit: how about put `resetOffsets` as the first condition so that if it's false, we would skip the synchronized function (not sure if JIT would really be able to optimize this way)?

Reply:
Will do

Reply:
We kind of assume that an application will not be doing heavy/frequent #removeNamedTopology calls, if it turns out that users want to be able to add and remove many topologies at a high rate then we can come back and try to optimize this -- it just doesn't seem to make much sense for an application to have such high turnover of topologies, this feature is generally speaking more targeted at providing a relatively stable application the ability to update its topology as needed, not for high volumes of transient topologies

Reply:
Oh mm actually this change is what's causing the tests to hang as it breaks the fix -- we actually need to ensure that we check the CREATED state and complete the future if so. But while I still stand by my comments above, ie that trying to avoid entering a synchronized block when adding or removing a topology is probably premature optimization, I actually did look over the class and believe we can make this work without synchronizing this particular method.

(However we do still need to synchronize on `start`, for several reasons)

Reply:
Thanks! I overlooked on its side effects..